docs/GettingStarted.md

```diff
@@ -311,6 +312,21 @@ current_root_span = Datadog.tracer.active_root_span
 current_root_span.set_tag('my_tag', 'my_value') unless current_root_span.nil?
 ```
 
+### Tracing threads
+
+`Datadog.tracer` keeps track of spans using a thread-local context. When new ruby thread is started, default behaviour is that `Datadog.tracer` creates new context for the new thread. As a consequence, the first span started inside of this new thread would be by default root span of the thread-local context.
+
+To connect the thread span as a child of the span in which the thread was started, start a new span inside of the thread and set its `parent` to the active span from outside of the thread. You can also use `child_of` option of `Datadog.tracer.trace` and `Datadog.tracer.start_span` methods.
+
+```ruby 
+parent_span = Datadog.tracer.active_span
+Thread.new do
+  Datadog.tracer.trace("thread", child_of: parent_span) do |thread_span|
```

Comment:
interesting. is this something added recently? can you tell me how does this work?

Reply:
I can't see `thread_local_parent_span` being implemented in the Tracer class :/

Reply:
Hey @DocX, I'm sorry, but my comment above had partial pseudo Ruby.

When setting the parent using `child_of`, only that single span gets associated with the parent thread. If we perform other instrumentation while inside that block, e.g.:
```ruby
parent_span = Datadog.tracer.active_span
Thread.new do
  Datadog.tracer.trace("thread", child_of: parent_span) do |thread_span|
     Service.send_http_request
  end
end
```

In this example, if `Service.send_http_request` generates spans, they won't be a child of `Datadog.tracer.trace("thread", child_of: parent_span)`, as the thread-local context (`Datadog.tracer.provider.context`) has not been changed.

Changing the thread-local context will work, but it will cause the issue described by @delner: because both the original thread and the new thread share the same context, the traces will be mixed together arbitrarily.

We currently don't have a concept that is able to handle multiple threads sharing the same trace, while maintaining two separate stack traces.

So, performing `Datadog.tracer.trace("thread", child_of: parent_span)` will definitely work for that single span, but not nested spans (unless we explicitly continue the `child_of` chain). Do you think that provides enough of the value you were looking for?

Reply:
Ok I can see it now. I've just checked the `Tracer` code and yes, I think there is difference between passing the child of as parameter to the new trace `trace(child_of: parent)` and setting the parent inside the block, after the span was created.

**setting parent inside of the block:**

So we in our production code still use this code:

```rb
Thread.new do 
  Datadog.tracer.trace("thread") do |span|
    span.parent = parent_span
    # ... nested spans here
  end
end
```

And this works perfectly:

<img width="843" alt="Screenshot 2020-09-11 at 09 31 00" src="https://user-images.githubusercontent.com/132277/92885601-3214c780-f413-11ea-8f2f-91b79e8ccd0f.png">

I now understand that this is because when we call the `trace("thread")` without `child_of` the tracer actually creates the new thread local context for that span first via

https://github.com/DataDog/dd-trace-rb/blob/0574335134aad3eb57cb6515f73943bc1afd6b60/lib/ddtrace/tracer.rb#L262

which calls this which creates the new thread local instance of context:

https://github.com/DataDog/dd-trace-rb/blob/0574335134aad3eb57cb6515f73943bc1afd6b60/lib/ddtrace/tracer.rb#L64

Allowing all nested spans inside of that thread to correctly pick this "root thread" span as parent.

**child_of: option**

In this case the very same line will not evaluate and will not actually create the thread local context:

Reply:
Ha, I didn't think of this scenario, thank you for that!
The screenshots look as I'd expect for an async trace, so that's great to see.

You probably remember that `:child_of` not completely functional before you fixed it a few months ago, so I think it's ok to consider changing its API semantics if we need to.

I agree with you that there should be only one way to accomplish this async tracing, and it should be as simple as possible.

I'm going to bring this to our team and get back to you, but I believe we'll make some code changes before these documentation changes get merged :)